### ISSUE TITLE
🐛 Bug Report: Increased memory usage with devtools plugin (catalog-unprocessed-entities)

### ISSUE THREAD

**Issue by sathish-soundarapandian** [backstage/backstage#22783]
### 📜 Description

Recently we integrated dev-tools plugin in our application to surface out catalog unprocessed entities in the frontend.
While things work fine, we see increased memory usage with these changes. Running the application locally we were not able to catch this.

But in upper development environment (Running the app in Kubernetes - memory limit was set to 500 MB), app crashed at startup itself. Then we increase the memory limit to 1 GB for it to get going.

Post that we used this tool (https://github.com/bnoordhuis/node-heapdump) to heap-dump with and without those changes. Heap dump was taken right after the startup with same setup on both occasions.
```
With Devtools plugin changes -
	Time taken for heapdump snapshot - 2 minute 23 seconds
	Heapdump snapshot size - 173 MB

Before Devtools plugin changes -
	Time taken for heapdump snapshot - 1 minute 32 seconds
	Heapdump snapshot size - 100 MB
```
Above data is consistent on all runs.

Is this devtools plugin "catalog-unprocessed-entities" suppose to take more memory in heap ?
Going through the plugin, it is just surfacing out the unprocessed entities from database (refresh_state, final_entities)

### 👍 Expected behavior

We shouldn't see increased memory usage with this plugin.

### 👎 Actual Behavior with Screenshots

Memory pattern -
```
With Devtools plugin changes -
	Time taken for heapdump snapshot - 2 minute 23 seconds
	Heapdump snapshot size - 173 MB

Before Devtools plugin changes -
	Time taken for heapdump snapshot - 1 minute 32 seconds
	Heapdump snapshot size - 100 MB
```

### 👟 Reproduction steps

Integrating dev tools catalog unprocessed entities plugin in an existing application.

### 📃 Provide the context for the Bug.

We wanted to make use of this dev tools plugin for surfacing out catalog unprocessed entities in the UI. But even with limited entities (< 25 catalog entities), we are observing increased memory usage in the application

### 🖥️ Your Environment

```
OS:   Darwin 23.2.0 - darwin/arm64
node: v16.20.0
yarn: 3.2.3
cli:  0.22.12 (installed)
backstage:  1.17.5

Dependencies:
  @backstage/app-defaults                                          1.4.4
  @backstage/backend-app-api                                       0.5.7, 0.5.9
  @backstage/backend-common                                        0.19.8, 0.19.9, 0.20.0
  @backstage/backend-dev-utils                                     0.1.2
  @backstage/backend-openapi-utils                                 0.0.3, 0.0.5
  @backstage/backend-plugin-api                                    0.6.6, 0.6.8
  @backstage/backend-tasks                                         0.5.11, 0.5.13
  @backstage/catalog-client                                        1.4.5, 1.4.6, 1.5.1
  @backstage/catalog-model                                         1.2.0, 1.4.3
  @backstage/cli-common                                            0.1.13
  @backstage/cli-node                                              0.1.5, 0.2.1
  @backstage/cli                                                   0.22.14
  @backstage/config-loader                                         1.5.1, 1.5.3, 1.6.0
  @backstage/config                                                1.1.1
  @backstage/core-app-api                                          1.11.0, 1.11.1
  @backstage/core-components                                       0.12.4, 0.11.2, 0.12.5, 0.13.6, 0.13.8, 0.13.9
  @backstage/core-plugin-api                                       1.0.6, 1.7.0, 1.8.0, 1.8.1
  @backstage/dev-utils                                             1.0.22
  @backstage/errors                                                1.2.3
  @backstage/eslint-plugin                                         0.1.3
  @backstage/frontend-app-api                                      0.2.0
  @backstage/frontend-plugin-api                                   0.2.0, 0.3.0
  @backstage/integration-aws-node                                  0.1.7, 0.1.8
  @backstage/integration-react                                     1.1.20, 1.1.21
  @backstage/integration                                           1.7.1, 1.7.2, 1.8.0
  @backstage/plugin-api-docs                                       0.9.12
  @backstage/plugin-app-backend                                    0.3.54
  @backstage/plugin-app-node                                       0.1.6
  @backstage/plugin-auth-backend                                   0.18.8
  @backstage/plugin-auth-node                                      0.2.19, 0.3.2, 0.4.0, 0.4.2
  @backstage/plugin-catalog-backend-module-github                  0.3.7
  @backstage/plugin-catalog-backend-module-scaffolder-entity-model 0.1.3, 0.1.5
  @backstage/plugin-catalog-backend-module-unprocessed             0.3.2
  @backstage/plugin-catalog-backend                                1.14.0
  @backstage/plugin-catalog-common                                 1.0.17, 1.0.18, 1.0.19
  @backstage/plugin-catalog-graph                                  0.2.37
  @backstage/plugin-catalog-import                                 0.9.13
  @backstage/plugin-catalog-node                                   1.4.7, 1.6.0
  @backstage/plugin-catalog-react                                  1.2.1, 1.8.5, 1.9.0
  @backstage/plugin-catalog-unprocessed-entities                   0.1.3
  @backstage/plugin-catalog                                        1.14.0
  @backstage/plugin-circleci                                       0.3.25
  @backstage/plugin-devtools-common                                0.1.7
  @backstage/plugin-devtools                                       0.1.7
  @backstage/plugin-dynatrace                                      8.0.1
  @backstage/plugin-events-node                                    0.2.15
  @backstage/plugin-firehydrant                                    0.2.9
  @backstage/plugin-github-actions                                 0.6.6
  @backstage/plugin-github-pull-requests-board                     0.1.19
  @backstage/plugin-graphiql                                       0.2.55
  @backstage/plugin-home-react                                     0.1.4, 0.1.5
  @backstage/plugin-home                                           0.5.10
  @backstage/plugin-jenkins-backend                                0.2.8
  @backstage/plugin-jenkins-common                                 0.1.20
  @backstage/plugin-jenkins                                        0.8.6
  @backstage/plugin-kubernetes-backend                             0.11.5
  @backstage/plugin-kubernetes-common                              0.6.6
  @backstage/plugin-kubernetes                                     0.10.3
  @backstage/plugin-org                                            0.6.15
  @backstage/plugin-pagerduty                                      0.6.6
  @backstage/plugin-permission-backend                             0.5.29
  @backstage/plugin-permission-common                              0.7.9, 0.7.10, 0.7.11
  @backstage/plugin-permission-node                                0.7.17, 0.7.19
  @backstage/plugin-permission-react                               0.4.16, 0.4.17, 0.4.18
  @backstage/plugin-proxy-backend                                  0.3.2
  @backstage/plugin-scaffolder-backend-module-azure                0.1.0
  @backstage/plugin-scaffolder-backend-module-bitbucket            0.1.0
  @backstage/plugin-scaffolder-backend-module-gerrit               0.1.0
  @backstage/plugin-scaffolder-backend-module-github               0.1.0
  @backstage/plugin-scaffolder-backend-module-gitlab               0.2.11
  @backstage/plugin-scaffolder-backend                             1.18.1, 1.19.2
  @backstage/plugin-scaffolder-common                              1.4.2, 1.4.4
  @backstage/plugin-scaffolder-node                                0.2.7, 0.2.9
  @backstage/plugin-scaffolder-react                               1.5.6
  @backstage/plugin-scaffolder                                     1.15.1
  @backstage/plugin-search-backend-module-catalog                  0.1.10
  @backstage/plugin-search-backend-module-pg                       0.5.15
  @backstage/plugin-search-backend-module-techdocs                 0.1.10
  @backstage/plugin-search-backend-node                            1.2.10
  @backstage/plugin-search-backend                                 1.4.6
  @backstage/plugin-search-common                                  1.2.7, 1.2.8, 1.2.9
  @backstage/plugin-search-react                                   1.7.1
  @backstage/plugin-search                                         1.4.1
  @backstage/plugin-sonarqube-backend                              0.2.8
  @backstage/plugin-sonarqube-react                                0.1.9
  @backstage/plugin-sonarqube                                      0.7.7
  @backstage/plugin-techdocs-backend                               1.8.0
  @backstage/plugin-techdocs-module-addons-contrib                 1.0.16
  @backstage/plugin-techdocs-node                                  1.9.0
  @backstage/plugin-techdocs-react                                 1.1.9, 1.1.12
  @backstage/plugin-techdocs                                       1.6.6
  @backstage/plugin-todo-backend                                   0.2.2
  @backstage/plugin-todo                                           0.2.29
  @backstage/plugin-user-settings                                  0.7.11
  @backstage/release-manifests                                     0.0.10, 0.0.9
  @backstage/test-utils                                            1.4.4
  @backstage/theme                                                 0.2.19, 0.4.3, 0.4.4, 0.5.0
  @backstage/types                                                 1.1.1
  @backstage/version-bridge                                        1.0.6, 1.0.7
```

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

No, I don't have time to work on this right now

**Comment by sathish-soundarapandian** [comment]
Some details from the snapshots taken -
with represents snapshot taken with dev tools changes and without represents the heap dump snapshot without dev tools changes.

<img width="2056" alt="Screenshot 2024-02-07 at 12 17 05 PM" src="https://github.com/backstage/backstage/assets/117059391/729bd777-aabf-4980-a4e1-f2d0e84907ef">
<img width="2056" alt="Screenshot 2024-02-07 at 12 17 07 PM" src="https://github.com/backstage/backstage/assets/117059391/e12e520b-4b58-4473-91ad-6a98db973bf1">
<img width="2056" alt="Screenshot 2024-02-07 at 12 18 22 PM" src="https://github.com/backstage/backstage/assets/117059391/e627e94a-0bc6-467b-a096-cc9b0c173083">

**Comment by awanlin** [comment]
Hi @sathish-soundarapandian, something worth looking into is running `@backstage/plugin-catalog-unprocessed-entities` on it's own and not part of DevTools. `@backstage/plugin-catalog-unprocessed-entities` is actually a stand-along plugin and DevTools just happens to be a nice easy container you can use to include it in. Due to this I don't think DevTools would be the culprit and perhaps this is something in the `@backstage/plugin-catalog-backend-module-unprocessed` plugin which is the backend for the features seen in `@backstage/plugin-catalog-unprocessed-entities`

**Comment by awanlin** [comment]
@kurtaking you have this same setup, have you had any memory issues with it? No rush, just curious

**Comment by kurtaking** [comment]
The goal is to add the delete functionality and then migrate once released.

**Comment by sathish-soundarapandian** [comment]
Hi @awanlin Haven't tried it out yet. Will check and get back here.

**Comment by mikeyhc** [comment]
Hey @sathish-soundarapandian, out of curiosity were these profiling metrics gathered from the front or backend as you mention both kubernetes and the UI

**Comment by sathish-soundarapandian** [comment]
Hi @mikeyhc These metrics were gathered from the backend. Heapdump snapshot was taken from the pod where backstage application was running.
Just used chrome memory inspect to visualize the generated heapdump

**Comment by mikeyhc** [comment]
Hey @sathish-soundarapandian, in that case this behavior is expected. In addition to loading the plugin itself, as a debugging tool it stores a lot of data and seeing 200MB of additional memory usage isn't unusual. When running a development version of the catalog we generally recommend a 4GB memory limit to allow for usage of these kinds of tools.

**Comment by sathish-soundarapandian** [comment]
We just upgraded from backstage 1.17.5 to 1.21.1 and started using jemalloc as the default memory allocator for the process. With that we saw overall improvements -

1. App startup time reduced from 9 - 10 seconds to 3 - 4 seconds.
2. No abnormal memory spikes / linear growth. Looks like the heap memory was always fine, it's just we had native memory issues due to memory fragmentation and once we started using jemalloc as the default memory allocator for the process, it is now well in control.
3. Time taken to generate a heap dump came down to 30 seconds (Previously it was going over 3 minutes). Moreover previously after taking heap dump the memory wouldn't release, that's not the case anymore with new set of changes.
